File: 8.ApproximatePatternMatching/ApproximatePatternMatching.go
package main

import "fmt"

func HammingDistance(a, b string) int {
	if len(a) != len(b) {
		fmt.Println("Strings must have the same size")
		return 1
	}
	dist := 0
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			dist++
		}
	}
	return dist
}

func ApproximatePatternMatching(Pattern, Text string, d int) []int {
	var indices []int
	var k int
	for i := 0; i < len(Text)-len(Pattern)+1; i++ {
		k = HammingDistance(Pattern, Text[i:i+len(Pattern)])
		if k <= d {
			indices = append(indices, i)
		}
	}
	return indices
}

func main() {
	genome := "CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAAT"
	kmer := "ATTCTGGA"
	d := 3
	var ind []int
	ind = ApproximatePatternMatching(kmer, genome, d)
	fmt.Println(ind)
}

File: 8.ApproximatePatternMatching/ApproximatePatternMatching.py


def hamming_distance(p: str, q: str):
    return sum(1 for x, y in zip(p, q) if x != y)


def approximate_pattern_matching(pattern: str, text: str, d: int):
    indices = []
    for i in range(len(text) - len(pattern) + 1):
        k = hamming_distance(text[i:i+len(pattern)], pattern)
        if k <= d:
            indices.append(i)
    return indices


def main():
    genome = "CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAAT"
    kmer = "ATTCTGGA"
    d = 3
    answer = approximate_pattern_matching(kmer, genome, d)
    assert answer == [6, 7, 26, 27]


if __name__ == '__main__':
    main()
